## Load in the data

The dataset is from house sales in King County, the region where the city of Seattle, WA is located.  
https://www.kaggle.com/harlfoxem/housesalesprediction

In [1]:
import pandas as pd

In [2]:
!wget https://www.dropbox.com/s/l6zj8mjqgdm1mne/kc_house_data.csv -q

In [3]:
df = pd.read_csv('kc_house_data.csv')

In [5]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view     ...      grade  sqft_above  \
0      5650     1.0           0     0     ...          7        1180   
1      7242     2.0           0     0     ...          7        2170   
2     10000     1.0           0     0     ...          6         770   
3      5000     1.0           0     0     ...          7        1050   
4      8080     1.0           0     0     ...          8        1680   

   sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0              0      1955             0    98178  47.5112 -122.257   
1            400      1951          1991    98125  47.7210 -122.319   
2              0      1933             0    98028  47.7379 -122.233   
3            910      1965             0    98136  47.5208 -122.393   
4              0      1987             0    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503  

[5 rows x 21 columns]

In [6]:
df.shape

(21613, 21)

In [7]:
m = df.shape[0] # number of training examples
m

21613

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
id               21613 non-null int64
date             21613 non-null object
price            21613 non-null float64
bedrooms         21613 non-null int64
bathrooms        21613 non-null float64
sqft_living      21613 non-null int64
sqft_lot         21613 non-null int64
floors           21613 non-null float64
waterfront       21613 non-null int64
view             21613 non-null int64
condition        21613 non-null int64
grade            21613 non-null int64
sqft_above       21613 non-null int64
sqft_basement    21613 non-null int64
yr_built         21613 non-null int64
yr_renovated     21613 non-null int64
zipcode          21613 non-null int64
lat              21613 non-null float64
long             21613 non-null float64
sqft_living15    21613 non-null int64
sqft_lot15       21613 non-null int64
dtypes: float64(5), int64(15), object(1)
memory usage: 3.5+ MB


In [9]:
df.isnull().sum()

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

The dataset contains information (21 features, including the price) related to 21613 houses.  
Our target variable (i.e., what we want to predict when a new house gets on sale) is the price.

In [10]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21613.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.390691     291.509045   1971.005136     84.402258   
std        1.175459    828.090978     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

## Baseline: the simplest model

Now let's compute the loss in the case of the simplest model: a fixed price equal to the average of historic prices, independently on house size, rooms, location, ...

In [11]:
# Let's compute the mean of the House Prices in King County 
y = df['price'] # extract the price column

In [12]:
avg_price = y.mean() # this is our baseline
print ("average price: ${:.0f} ".format(avg_price))

average price: $540088 


In [13]:
ExamplePrice = y[0]
ExamplePrice

221900.0

The predictions are very easy to calculate, just the baseline value:

In [14]:
def get_baseline_predictions():
       # Simplest version: return the baseline as predicted values
    predicted_values = avg_price 
    return predicted_values


Example:

In [15]:
my_house_size = 2500
estimated_price = get_baseline_predictions()
print ("The estimated price for a house with {} squared feet is {:.0f}".format(my_house_size, estimated_price))

The estimated price for a house with 2500 squared feet is 540088


The estimated price for the example house will still be around 540K, wile the real value is around 222K. Quite an error!

## Measures of loss 

There are several way of implementing the loss, I use the squared error here.  

$L = [y - f(X)]^2$

$J = SQRT(Sum(L)/M)$

In [16]:
import numpy as np
def get_loss(yhat, target):
    """
    Arguments:
    yhat -- vector of size m (predicted labels)
    target -- vector of size m (true labels)
    
    Returns:
    loss -- the value of the L2 loss function
    """
      # compute the residuals (since we are squaring it doesn't matter 
      #   which order you subtract)
      # np.dot will square the residuals and add them up
    cost = np.dot((target - yhat), (target - yhat))/len(yhat)

    return(np.sqrt(cost))

In [17]:
#predict the sale price is the mean.
y = df['price']
mean = sum(y)/len(y)
print("Model Errors for Mean")
print("RMSE = {:.0f}".format(get_loss(y,mean)))

Model Errors for Mean
RMSE = 367119


In [18]:
model_costs = []
model_costs.append(get_loss(y,mean))
model_costs

[367118.7031813723]

To better see the value of the cost function we use also the RMSE, the Root Mean Square Deviation.  
Basically the average of the losses, rooted.

In [19]:
from sklearn import linear_model
def model_behaviour(df, features, target):
  
    y = df[target]
    X = df[features]
    
    model = linear_model.LinearRegression()
    
    model.fit(X, y)
    
    y_hat = model.predict(X)
    
    cost = get_loss(y_hat, y)
    
    C0 = model.intercept_
    Cs = model.coef_  
    equation = str(np.round(C0))
    for i in range(len( Cs)):
       equation += " + " + str(np.round(Cs[i])) + "*" + features[i]
    print ("y = ", equation)
    print("RMSE = {:.0f}".format(cost))

    return cost

    
    

In [20]:
cost = model_behaviour(df,["sqft_living"],"price")
model_costs.append(cost)
model_costs

y =  -43581.0 + 281.0*sqft_living
RMSE = 261441


[367118.7031813723, 261440.79030067177]

As you can see, it is quite high error, especially related to the average selling price.

Now, we can look at how training error behaves as model complexity increases.

## Learning a multiple regression model

We can add more features to the model, for example the number of bedrooms and bathrooms.

In [21]:
more_features = ['sqft_living', 'bedrooms', 'bathrooms']

In [22]:
cost = model_behaviour(df,df.columns.drop(['price','date']),"price")
model_costs.append(cost)
model_costs

y =  6938892.0 + -0.0*id + -35763.0*bedrooms + 41156.0*bathrooms + 110.0*sqft_living + 0.0*sqft_lot + 6765.0*floors + 582652.0*waterfront + 53031.0*view + 26255.0*condition + 96007.0*grade + 71.0*sqft_above + 40.0*sqft_basement + -2622.0*yr_built + 20.0*yr_renovated + -582.0*zipcode + 602355.0*lat + -212793.0*long + 22.0*sqft_living15 + -0.0*sqft_lot15
RMSE = 201131


[367118.7031813723, 261440.79030067177, 201130.68126357702]

In [23]:
cost = model_behaviour(df,more_features,"price")
model_costs.append(cost)
model_costs

y =  74847.0 + 309.0*sqft_living + -57861.0*bedrooms + 7933.0*bathrooms
RMSE = 257790


[367118.7031813723, 261440.79030067177, 201130.68126357702, 257789.5411485218]

We can learn a multiple regression model predicting 'price' based on the above features on the data with the following code:

# Create some new features

Although we often think of multiple regression as including multiple different features (e.g. # of bedrooms, squarefeet, and # of bathrooms) but we can also consider transformations of existing features e.g. the log of the squarefeet or even "interaction" features such as the product of bedrooms and bathrooms.

In [24]:
from math import log

Next we create the following new features as column :
* bedrooms_squared = bedrooms\*bedrooms
* bed_bath_rooms = bedrooms\*bathrooms
* log_sqft_living = log(sqft_living)
* lat_plus_long = lat + long 
* more polynomial features: bedrooms ^ 4, bathrooms ^ 7, size ^ 3


In [25]:
df['bedrooms_squared'] = df['bedrooms'].apply(lambda x: x**2)

In [26]:
df['bed_bath_rooms'] = df['bedrooms'] * df.bathrooms

In [27]:
df['log_sqft_living'] = df['sqft_living'].apply(lambda x: log(x))

In [28]:
df['lat_plus_long'] = df['lat'] + df.long

In [29]:
df['bedrooms_4'] = df['bedrooms'].apply(lambda x: x**4)

In [30]:
df['bathrooms_7'] = df['bathrooms'].apply(lambda x: x**7)

In [31]:
df['size_3'] = df['sqft_living'].apply(lambda x: x**3)

In [32]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view     ...          long  sqft_living15  \
0      5650     1.0           0     0     ...      -122.257           1340   
1      7242     2.0           0     0     ...      -122.319           1690   
2     10000     1.0           0     0     ...      -122.233           2720   
3      5000     1.0           0     0     ...      -122.393           1360   
4      8080     1.0           0     0     ...      -122.045           1800   

   sqft_lot15  bedrooms_squared  bed_bath_rooms  log_sqft_living  \
0        5650                 9            3.00         7.073270   
1        7639                 9            6.75         7.851661   
2        8062                 4            2.00         6.646391   
3        5000                16           12.00         7.580700   
4        7503                 9            6.00         7.426549   

   lat_plus_long  bedrooms_4  bathrooms_7       size_3  
0       -74.7458          81      1.00000   1643032000  
1       -74.5980          81    291.92926  16974593000  
2       -74.4951          16      1.00000    456533000  
3       -74.8722         256   2187.00000   7529536000  
4       -74.4282          81    128.00000   4741632000  

[5 rows x 28 columns]

* Squaring bedrooms will increase the separation between not many bedrooms (e.g. 1) and lots of bedrooms (e.g. 4) since 1^2 = 1 but 4^2 = 16. Consequently this feature will mostly affect houses with many bedrooms.
* bedrooms times bathrooms gives what's called an "interaction" feature. It is large when *both* of them are large.
* Taking the log of squarefeet has the effect of bringing large values closer together and spreading out small values.
* Adding latitude to longitude is totally non-sensical but we will do it anyway (you'll see why)

## Learning Multiple Models

Now we will learn the weights for five (nested) models for predicting house prices. The first model will have the fewest features, the second model will add more features and so on:

In [33]:
model_1_features = ['sqft_living', 'bedrooms', 'bathrooms', 'lat', 'long', 'sqft_lot', 'floors']
model_2_features = model_1_features + ['log_sqft_living', 'bedrooms_squared', 'bed_bath_rooms']
model_3_features = model_2_features + ['lat_plus_long']
model_4_features = model_3_features + ['bedrooms_4', 'bathrooms_7']
model_5_features = model_4_features + ['size_3']

Now that we have the features, we learn the weights for the five different models for predicting target = 'price' using and look at the value of the weights/coefficients:

In [34]:
cost = model_behaviour(df,model_1_features,"price")
model_costs.append(cost)


y =  -69873110.0 + 307.0*sqft_living + -53737.0*bedrooms + 19263.0*bathrooms + 659920.0*lat + -315325.0*long + 0.0*sqft_lot + -5770.0*floors
RMSE = 234905


In [35]:
cost = model_behaviour(df,model_2_features,"price")
model_costs.append(cost)


y =  -61776267.0 + 522.0*sqft_living + -5831.0*bedrooms + 91387.0*bathrooms + 655377.0*lat + -279014.0*long + -0.0*sqft_lot + -1641.0*floors + -543969.0*log_sqft_living + 766.0*bedrooms_squared + -14998.0*bed_bath_rooms
RMSE = 227350


In [36]:
cost = model_behaviour(df,model_3_features,"price")
model_costs.append(cost)


y =  -61776267.0 + 522.0*sqft_living + -5831.0*bedrooms + 91387.0*bathrooms + 529923.0*lat + -404469.0*long + -0.0*sqft_lot + -1641.0*floors + -543969.0*log_sqft_living + 766.0*bedrooms_squared + -14998.0*bed_bath_rooms + 125454.0*lat_plus_long
RMSE = 227350


In [37]:
cost = model_behaviour(df,model_4_features,"price")
model_costs.append(cost)


y =  -62393515.0 + 507.0*sqft_living + 37976.0*bedrooms + 73081.0*bathrooms + 533075.0*lat + -406804.0*long + -0.0*sqft_lot + 442.0*floors + -514539.0*log_sqft_living + -6644.0*bedrooms_squared + -11263.0*bed_bath_rooms + 126271.0*lat_plus_long + 6.0*bedrooms_4 + 1.0*bathrooms_7
RMSE = 226783


In [38]:
cost = model_behaviour(df,model_5_features,"price")
model_costs.append(cost)


y =  -62848726.0 + 438.0*sqft_living + 27657.0*bedrooms + 59701.0*bathrooms + 533192.0*lat + -406016.0*long + -0.0*sqft_lot + 760.0*floors + -412311.0*log_sqft_living + -6572.0*bedrooms_squared + -6800.0*bed_bath_rooms + 127175.0*lat_plus_long + 6.0*bedrooms_4 + -0.0*bathrooms_7 + 0.0*size_3
RMSE = 226533


In [39]:
model_1 = linear_model.LinearRegression()
model_1.fit(df[model_1_features], y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [40]:
model_2 = linear_model.LinearRegression()
model_2.fit(df[model_2_features], y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [41]:
model_3 = linear_model.LinearRegression()
model_3.fit(df[model_3_features], y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [42]:
model_4 = linear_model.LinearRegression()
model_4.fit(df[model_4_features], y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [43]:
model_5 = linear_model.LinearRegression()
model_5.fit(df[model_5_features], y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [44]:
# You can examine/extract each model's coefficients, for example:
print (model_1.coef_)
print (model_2.coef_)

[ 3.06930393e+02 -5.37365974e+04  1.92633985e+04  6.59919715e+05
 -3.15325117e+05  6.01533077e-02 -5.77002641e+03]
[ 5.22290856e+02 -5.83083200e+03  9.13874631e+04  6.55377075e+05
 -2.79014318e+05 -4.38190098e-03 -1.64123192e+03 -5.43968799e+05
  7.65579830e+02 -1.49981031e+04]


Interesting: in the previous model the weight coefficient for the size lot was positive but now in the model_2 is negative.  
This is an effect of adding the feature logging the size.

## Comparing multiple models

Now that you've learned three models and extracted the model weights we want to evaluate which model is best.

We can use the loss function from earlier to compute the RSS on training data for each of the models.

In [45]:
# Compute the RSS for each of the models:
print (get_loss(model_1.predict(df[model_1_features]), y))
print (get_loss(model_2.predict(df[model_2_features]), y))
print (get_loss(model_3.predict(df[model_3_features]), y))
print (get_loss(model_4.predict(df[model_4_features]), y))
print (get_loss(model_5.predict(df[model_5_features]), y))

234904.62730622126
227349.81661198896
227349.81661198914
226782.75610270022
226532.7041181746


model_5 has the lowest RSS on the training data.  
The **most complex model**.

## The test error

Training error decreases quite significantly with model complexity. This is quite intuitive, because the model was fit on the training points and then as we increase the model complexity, we are better able to fit the training data points.

A natural question is whether a training error is a good measure of predictive performance?  

The issue is that the training error is overly optimistic and that's because the beta parameters were fit on the training data to minimise the residual sum of squares, which can often be related to the training error.
So, in general, having small training error does not imply having good predictive performance.

This takes us to something called test error (or out-of-sample error): we hold out some houses from the data set and we're putting these into what's called a test set.
And when we fit our models, we just fit our models on the training data set.  
But then when we go to assess our performance of that model we look at these test houses in the test dataset and these are hopefully serving as a proxy of everything out there in the world.

Bottom line, the test error is a (noisy) approximation of the true error.

## Split data into training and testing.
Let's see how can be applied to our example.  
First we split the data into a training set and a testing set using a function from *sklearn*, the *train_test_split()*.
We use a seed for reproducibility.   

In [46]:
from sklearn.model_selection import train_test_split
train_data,test_data = train_test_split(df, test_size=0.3, random_state=999)

In [47]:
train_data.head()

id             date     price  bedrooms  bathrooms  \
9318   4420600015  20141006T000000  571500.0         4       2.25   
15904  8691390980  20140902T000000  728000.0         4       2.50   
7599   3326059191  20140616T000000  410000.0         4       2.00   
14893  6817810190  20140701T000000  401000.0         3       2.00   
20508  7625703357  20150227T000000  394950.0         2       2.25   

       sqft_living  sqft_lot  floors  waterfront  view     ...          long  \
9318          2810     25990     2.0           0     0     ...      -122.293   
15904         3290      5951     2.0           0     0     ...      -121.976   
7599          1580      9581     1.0           0     0     ...      -122.165   
14893         1240     11172     1.0           0     0     ...      -122.037   
20508         1300      2104     2.0           0     0     ...      -122.388   

       sqft_living15  sqft_lot15  bedrooms_squared  bed_bath_rooms  \
9318            2020       16140                16             9.0   
15904           3240        6159                16            10.0   
7599            1580       10552                16             8.0   
14893           1330       14102                 9             6.0   
20508           1430        1850                 4             4.5   

       log_sqft_living  lat_plus_long  bedrooms_4  bathrooms_7       size_3  
9318          7.940940       -74.9937         256   291.929260  22188041000  
15904         8.098643       -74.3761         256   610.351562  35611289000  
7599          7.365180       -74.4730         256   128.000000   3944312000  
14893         7.122867       -74.4006          81   128.000000   1906624000  
20508         7.170120       -74.8403          16   291.929260   2197000000  

[5 rows x 28 columns]

In [48]:
train_data.shape

(15129, 28)

In [49]:
# test_data = pd.read_csv('kc_house_test_data.csv', dtype=dtype_dict)
test_data.head()

id             date      price  bedrooms  bathrooms  \
10033  4440400125  20140508T000000   228000.0         4       1.75   
20870  3575305485  20140829T000000   409000.0         3       2.50   
21323  9402800005  20141028T000000  1500000.0         3       3.50   
12147  9407001600  20150423T000000   305000.0         3       1.75   
15145  8712100020  20150127T000000   600000.0         2       1.00   

       sqft_living  sqft_lot  floors  waterfront  view     ...          long  \
10033         2000      6120     1.0           0     0     ...      -122.258   
20870         1890      6500     2.0           0     0     ...      -122.058   
21323         3530      3610     2.0           0     0     ...      -122.339   
12147         1660     11500     1.0           0     0     ...      -121.774   
15145         1290      4636     1.0           0     0     ...      -122.301   

       sqft_living15  sqft_lot15  bedrooms_squared  bed_bath_rooms  \
10033           1880        6120                16            7.00   
20870           2340        7500                 9            7.50   
21323           1780        3610                 9           10.50   
12147           1250       11000                 9            5.25   
15145           1940        4635                 4            2.00   

       log_sqft_living  lat_plus_long  bedrooms_4  bathrooms_7       size_3  
10033         7.600902       -74.7545         256    50.265076   8000000000  
20870         7.544332       -74.4355          81   610.351562   6751269000  
21323         8.169053       -74.6533          81  6433.929688  43986977000  
12147         7.414573       -74.3267          81    50.265076   4574296000  
15145         7.162397       -74.6617          16     1.000000   2146689000  

[5 rows x 28 columns]

In [50]:
test_data.shape

(6484, 28)

In this case the testing set will be the 30% (therefore the training set is 70% of the original data)

In [51]:
train_y = train_data.price # extract the price column
test_y = test_data.price

Retrain the models on training data only:

In [52]:
model_1.fit(train_data[model_1_features], train_y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [53]:
model_2.fit(train_data[model_2_features], train_y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [54]:
model_3.fit(train_data[model_3_features], train_y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [55]:
model_4.fit(train_data[model_4_features], train_y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [56]:
model_5.fit(train_data[model_5_features], train_y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None,
         normalize=False)

In [57]:
# Compute the RSS on TRAINING data for each of the models
print (get_loss(model_1.predict(train_data[model_1_features]), train_y))
print (get_loss(model_2.predict(train_data[model_2_features]), train_y))
print (get_loss(model_3.predict(train_data[model_3_features]), train_y))
print (get_loss(model_4.predict(train_data[model_4_features]), train_y))
print (get_loss(model_5.predict(train_data[model_5_features]), train_y))

238423.73623346974
228417.37486543707
228417.3748654373
226068.48002941316
222660.59332007836


Now compute the RSS on TEST data for each of the models.

In [58]:
# Compute the RSS on TESTING data for each of the three models and record the values:
print (get_loss(model_1.predict(test_data[model_1_features]), test_y))
print (get_loss(model_2.predict(test_data[model_2_features]), test_y))
print (get_loss(model_3.predict(test_data[model_3_features]), test_y))
print (get_loss(model_4.predict(test_data[model_4_features]), test_y))
print (get_loss(model_5.predict(test_data[model_5_features]), test_y))

226857.32965286236
226001.99527230184
226001.99527230565
232069.83536944428
250224.65391678546


The most complex model has the lowest error on the training data, but since that model has a non-sensical feature, it performs less well on the test data.

![Training error vs. model complexity](../images/models.png)